src/mscorlib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -348,332 +257,450 @@ internal ICollection<TKey> Keys
-                    if (_invalid)
+                    for (int entriesIndex = _buckets[bucket]; entriesIndex != -1; entriesIndex = _entries[entriesIndex].Next)
                     {
-                        return;
+                        if (_entries[entriesIndex].HashCode == -1)
+                        {
+                            continue;   // removed entry whose handle is awaiting condemnation by the finalizer.
+                        }
+
+                        object thisKey, thisValue;
+                        _entries[entriesIndex].depHnd.GetPrimaryAndSecondary(out thisKey, out thisValue);
+                        if (Equals(thisKey, key))
+                        {
+                            GC.KeepAlive(this); // ensure we don't get finalized while accessing DependentHandles.
+                            value = (TValue)thisValue;
+                            return (TKey)thisKey;
+                        }
                     }
-                    Entry[] entries = _entries;
+                }
+
+                GC.KeepAlive(this); // ensure we don't get finalized while accessing DependentHandles.
+                value = default(TValue);
+                return null;
+            }
 
-                    // Make sure anyone sneaking into the table post-resurrection
-                    // gets booted before they can damage the native handle table.
-                    _invalid = true;
-                    _entries = null;
-                    _buckets = null;
+            //----------------------------------------------------------------------------------------
+            // Precondition:
+            //     Must hold _lock.
+            //----------------------------------------------------------------------------------------
+            private void VerifyIntegrity()
+            {
+                if (_invalid)
+                {
+                    throw new InvalidOperationException(Environment.GetResourceString("CollectionCorrupted"));
+                }
+            }
 
+            //----------------------------------------------------------------------------------------
+            // Finalizer.
+            //----------------------------------------------------------------------------------------
+            ~Container()
+            {
+                // We're just freeing per-appdomain unmanaged handles here. If we're already shutting down the AD,
+                // don't bother.
+                //
+                // (Despite its name, Environment.HasShutdownStart also returns true if the current AD is finalizing.)
+
+                if (Environment.HasShutdownStarted || _invalid)
+                {
+                    return;
+                }
+
+                Entry[] entries = _entries;
+
+                // Make sure anyone sneaking into the table post-resurrection
+                // gets booted before they can damage the native handle table.
+                _invalid = true;
```

Comment:
There seems to be a small race here with lock-free get: Something could have resurrected the table and the lock-free get can be potentially accessing the entries being freed from other thread.